### ISSUE TITLE
Arrays of union types with nested fields with same name but different type causes "could not merge" error

### ISSUE THREAD

**Issue by rmarscher** [HoudiniGraphql/houdini#651]
### Describe the bug

With a schema that has an array with a union type, if there is a field with the same name with a different type in the union types, it causes a "could not merge" warning error during codegen and does not fill the data properly in the query results.

The same query works when run directly against the GraphQL server.

The "could not merge" error seems to be from a check when building the query selection code that makes sure that each type is the same.
https://github.com/HoudiniGraphql/houdini/blob/main/packages/houdini/src/codegen/generators/artifacts/utils.ts#L46-L56

I believe the deepMerge call that fails comes from here https://github.com/HoudiniGraphql/houdini/blob/main/packages/houdini/src/codegen/generators/artifacts/selection.ts#L49

The workaround is to alias the field names in the query fragments so no fragment in the union has the same field name with different types. But then you have to update all of your components to use the new field name and generally remember that you have aliased that field because it doesn't match the schema.

I'm interested to attempt to help create a fix, but it seems like a tricky problem.

### Severity

serious, but I can work around it

### Steps to Reproduce the Bug

Given a schema like this:

```graphql
interface Content {
  name: String!
}

type Page implements Content {
  id: ID!
  images: PageImages!
  name: String!
}

type PageImages {
  footer: String
  header: String
}

type Article implements Content {
  id: ID!
  images: ArticleImages!
  name: String!
}

type ArticleImages {
  author: String
  header: String
}

type Query {
  allContent: [Content!]!
}
```

And then this query:

```graphql
query AllContent {
    allContent {
        name
        ... on Page {
            images {
                header
                footer
            }
        }
        ... on Article {
            images {
                header
                author
            }
        }
    }
}
```

### Reproduction

https://github.com/rmarscher/houdinigraphql-fragment-merge-repro

**Comment by AlecAivazis** [comment]
Hey @rmarscher sorry to hear you're running into this. I probably won't find the time to look into this for a little bit but if you don't mind doing the exploration, I can provide whatever guidance you need here or on discord. 

At the surface, it sounds like the check in deep merge should only really apply to leaf nodes since those are the ones whose types must match. For the intermediate types, we should be able to get away with marking them as abstract and letting the runtime look at the API response for the type information (so it can compute the correct ID). At the moment, this [just happens for interfaces and unions](https://github.com/HoudiniGraphql/houdini/blob/main/packages/houdini/src/codegen/generators/artifacts/selection.ts#L195).

**Comment by rmarscher** [comment]
I started a branch with an e2e test that demonstrates the issue. I will chip away at a solution. Opened a PR in my fork just to be able to review the changes more easily for now - https://github.com/rmarscher/houdini/pull/1. Once I have a fix, I'll open a PR here.

**Comment by AlecAivazis** [comment]
Just a heads up @rmarscher - we ran into another issue that is a different manifestation of this bug. Thinking through it some more, the fix I outlined before won't address both situations and the "correct" approach is quite involved so I'm going to pick this up myself and try to get something fixed over the next few days. Once that's in place, I would love to verify that your e2e test is fixed too so I'd love it if you opened a PR with them. 

I'll ping  you once it's ready. Thanks for the patience!

**Comment by rmarscher** [comment]
Thanks a lot, AlecAivazis. I really appreciate your effort on all of this.

**Comment by AlecAivazis** [comment]
Well, that was a doozy - mind pulling that PR down and seeing if your issue is fixed?

**Comment by AlecAivazis** [comment]
Okay cool. If you could run generate with `-v` to get the stack trace as well as a few extra pieces of info I can hunt this down:


- The query that's failing
- relevant bits of the schema

**Comment by rmarscher** [comment]
OK got it. I have the `@paginate` directive on a relay connection that paginates over a union type.

It fails on line 246 of typeCheck.ts - https://github.com/HoudiniGraphql/houdini/blob/main/packages/houdini/src/codegen/validators/typeCheck.ts#L232-L248

The `type` var returned from connectionSelection is GraphQLUnionType which doesn't have a getFields() method.

A quick fix could be to only check the first type in the union. Insert this after line 241 -
```
if (!type.getFields) {
  type.getFields = () => type._types[0].getFields()
}
```

I am still running into the `A field with cursor pagination cannot go forwards an backwards simultaneously` error - but that's something else that we discussed in #666 .

If I remove all of the paginate directives, the codegen runs without errors! 🏆

**Comment by AlecAivazis** [comment]
> I have the @paginate directive on a relay connection that paginates over a union type.

wow talk about an edge case. I'll try to figure out something over the weekend 👍

**Comment by AlecAivazis** [comment]
Apart from the codegen running succesfully, do the actual values get written and read to the cache correctly?

**Comment by AlecAivazis** [comment]
@rmarscher this should be fixed in #753. Mind giving that a whirl?

**Comment by rmarscher** [comment]
Sorry I somehow missed your unexpectedly quick turnaround on this! I just removed the workaround we had in place and yes it's working!! 🎩 ✨✨✨

Our use case is search and listing service that paginates over multiple model types. Our models have common struct fields (images, content, metadata) that have different definitions per model despite having the same field name.
